wled00/tpm2record.cpp

```diff
@@ -0,0 +1,292 @@
+#include "wled.h"
+
+#ifdef WLED_USE_SD
+#define USED_STORAGE_FILESYSTEMS "SD, LittleFS"
+#include "SD_MMC.h"
+#else
+#define USED_STORAGE_FILESYSTEMS "LittleFS"
+#endif
+
+// This adds TPM2-file storing and playback capabilities to WLED.
+// 
+// What does it mean:
+//   You can now store short recorded animations on the ESP32 (in the ROM: no SD required) with a connected LED stripe.
+//
+// How to transfer the animation:
+//   WLED offers a web file manager under <IP_OF_WLED>/edit here you can upload a recorded *.TPM2 file
+//
+// How to create a recording:
+//   You can record with tools like Jinx
+//
+// How to load the animation:
+//   You can specify a preset to playback this recording with the following API command
+//   {"tpm2":{"file":"/record.tpm2"}}
+//
+//   You can specify a preset to playback this recording on a specific segment
+//   {"tpm2":{"file":"/record.tpm2", "seg":{"id":2}}
+//   {"tpm2":{"file":"/record.tpm2", "seg":2}
+//
+// How to trigger the animation:
+//   Presets can be triggered multiple interfaces e.g. via the json API, via the web interface or with a connected IR remote
+//
+// What next:
+//   - Playback from SD card is the next plan, here the length of the animation is less of a problem.
+//   - Playback and Recording of RGBW animations, as right now only RGB recordings are supported by WLED
+
+// reference spec of TPM2: https://gist.github.com/jblang/89e24e2655be6c463c56
+// - A packet contains any data of the TPM2 protocol, it
+//     starts with `TPM2_START` and ends with `TPM2_END`
+// - A frame contains the visual data (the LEDs color's) of one moment
+
+// --- CONSTANTS ---
+#define TPM2_START      0xC9
+#define TPM2_DATA_FRAME 0xDA
+#define TPM2_COMMAND    0xC0
+#define TPM2_END        0x36
+#define TPM2_RESPONSE   0xAA
+
+// infinite loop of animation
+#define RECORDING_REPEAT_LOOP -1
+
+// Default repeat count, when not specified by preset (-1=loop, 0=play once, 2=repeat two times)
+#define RECORDING_REPEAT_DEFAULT 0
+
+// --- Recording playback related ---
+File     recordingFile;
+uint8_t  colorData[4];
+uint8_t  colorChannels    = 3;
+uint32_t msFrameDelay     = 33; // time between frames
```

Comment:
@constant-flow - Is there a reason why we have a 33 ms delay between frames? If not, can we make this parameter driven, or part of the TPM2 file header somehow?

Use case that I'm thinking of is converting xLights export 'fseq' files (Falcon Player) into TPM2 format which typically uses 25ms, 40ms, or 100ms per frame (40 fps, 20 fps, and 10 fps respectively).

Reply:
@guardmedia Currently it's not bound to any limitation, it's simple for the sake of having a defined state. The TPM2 spec is rather lose on that end. The most recent specs are these from 2013 [LINK (sorry in 🇩🇪 )](https://www.ledstyles.de/content/index.php?attachment/28200-tpm2-specs-v1-0-2013-ger-pdf/)

There is a command `0x0C` that can be used to communicate the master speed. Unfortunately there is no concrete example what the payload should look like to set it to a specific frame rate, e.g. 60FPS

I was in contact with one of the original authors (Pesi) of the TPM2 protocol. He is not having the time nor the need as of now to continue the active development. He is still online in the **ledstyles** forum and would be okay if someone else continues the work on the "standard". My plan was to take their specifications and add clear examples and more precise rules how to set certain parameters, basically a new spec version. Next to framerate I'd also define color channel handling, e.g is it RGB or GRB or RGBW that the file is storing. The new specs i'd base of an [english translation](https://gist.github.com/jblang/89e24e2655be6c463c56) that already exists and publish it on Github. If Pesi likes the new specs, he might even promote it and host it under his reserved `tpm2.de` domain.

So I'm totally with u, we should have the framerate and other information packed into the file. I simply haven't had the time to work on the new specs with examples ... but I got the okay from both parties to alter what is there :) So maybe the WLED community can push (revive) the TPM2 format. (naming is a bit unlucky as it's conflicting with ISO/IEC 11889:2015)

A converter or player for fseq is definitely a cool and required feature for some more advanced / custom animations or use-cases.

Reply:
Thank you for the detailed response! Doing a quick search through the TPM2 spec, it seems that framerate is typically driven by the sending source, baud rate, or processing power. In all documentation I could find, the actual FPS was always determined by the author. For example, [Jinx!](http://www.live-leds.de/) uses a fixed framerate of 25fps or 20fps if using the **-p** switch.

From the Jinx! user manual (http://www.live-leds.de/download/190/) ...

Page 27
> _**Attention:** Jinx! will always produce 25fps. So be sure that your output device is able to handle it (e.g. serial
line with a too small baud rate), otherwise a frame skip will appear_

Page 35
> **Limit output frame rate to 20fps**
Additionally you can use the -p switch to limit the output frame rate to 20fps. This option is for debugging
purposes and not designed to use in a production state. The internal generators will work with 25fps
regardless off the -p switch.

Long story short, 33 ms delay (~30 fps) sounds like a reasonable rate. Once I have more time, I will look into implementing FSEQ directly instead of trying to re-write the TPM2 protocol.

Thank you for the great idea to add SD ROM support into WLED - looking forward to seeing this getting merged!

Reply:
Nice to see your implementation for the FSEQ format. Yeah I checked the "default" FPS but sorry 20/25 FPS is not really what i thought is usable 😂 . My recorder uses 30 FPS for now (I hope i can publish it as well).

The ideal solution is to set FPS via the headers in the recording, we all agree.

Let's see when our PR will be merged in 💍